File: tests/corpus/concept.js
// vyakarana stand-in -- not from vidya. See docs/adrs/0006.
// Re-sync when vidya adds a JS reference sample.

'use strict';

const STANDIN = true;
const NOTE = "vyakarana stand-in for JS; delete when vidya ships a sample.";

// ── Tokens ───────────────────────────────────────────────────────────

const TokenKind = Object.freeze({
    Number:  'number',
    Plus:    'plus',
    Minus:   'minus',
    Star:    'star',
    Slash:   'slash',
    LParen:  'lparen',
    RParen:  'rparen',
    Eof:     'eof',
});

class Token {
    constructor(kind, text, pos) {
        this.kind = kind;
        this.text = text;
        this.pos  = pos;
    }
    toString() {
        return `Token(${this.kind}, ${this.text}, ${this.pos})`;
    }
}

// ── Lexer ────────────────────────────────────────────────────────────

class Lexer {
    constructor(source) {
        this.source = source;
        this.pos = 0;
    }

    peek() {
        return this.pos < this.source.length ? this.source[this.pos] : null;
    }

    advance() {
        return this.pos < this.source.length ? this.source[this.pos++] : null;
    }

    skipWhitespace() {
        while (this.pos < this.source.length && /\s/.test(this.source[this.pos])) {
            this.pos++;
        }
    }

    next() {
        this.skipWhitespace();
        if (this.pos >= this.source.length) {
            return new Token(TokenKind.Eof, '', this.pos);
        }
        const start = this.pos;
        const ch = this.source[this.pos];

        if (/[0-9]/.test(ch)) {
            while (this.pos < this.source.length && /[0-9]/.test(this.source[this.pos])) {
                this.pos++;
            }
            return new Token(TokenKind.Number, this.source.slice(start, this.pos), start);
        }

        const ops = {
            '+': TokenKind.Plus,
            '-': TokenKind.Minus,
            '*': TokenKind.Star,
            '/': TokenKind.Slash,
            '(': TokenKind.LParen,
            ')': TokenKind.RParen,
        };
        if (ops[ch] !== undefined) {
            this.pos++;
            return new Token(ops[ch], ch, start);
        }

        throw new Error(`unexpected character '${ch}' at position ${start}`);
    }
}

// ── Pratt parser ─────────────────────────────────────────────────────

function parseExpr(lexer, minBp = 0) {
    let lhs;
    let tok = lexer.next();

    if (tok.kind === TokenKind.Number) {
        lhs = Number(tok.text);
    } else if (tok.kind === TokenKind.Minus) {
        const inner = parseExpr(lexer, 9);
        lhs = -inner;
    } else if (tok.kind === TokenKind.LParen) {
        lhs = parseExpr(lexer, 0);
        const close = lexer.next();
        if (close.kind !== TokenKind.RParen) {
            throw new Error(`expected ')' at ${close.pos}`);
        }
    } else {
        throw new Error(`expected expression at ${tok.pos}, got ${tok.kind}`);
    }

    for (;;) {
        const op = lexer.next();
        const bp = bindingPower(op.kind);
        if (bp === null || bp <= minBp) {
            lexer.pos -= (op.kind === TokenKind.Eof ? 0 : op.text.length);
            break;
        }
        const rhs = parseExpr(lexer, bp);
        lhs = applyOp(op.kind, lhs, rhs);
    }
    return lhs;
}

function bindingPower(kind) {
    switch (kind) {
        case TokenKind.Plus:
        case TokenKind.Minus:
            return 1;
        case TokenKind.Star:
        case TokenKind.Slash:
            return 3;
        default:
            return null;
    }
}

function applyOp(kind, lhs, rhs) {
    switch (kind) {
        case TokenKind.Plus:  return lhs + rhs;
        case TokenKind.Minus: return lhs - rhs;
        case TokenKind.Star:  return lhs * rhs;
        case TokenKind.Slash: return lhs / rhs;
        default: throw new Error(`unknown op ${kind}`);
    }
}

// ── Smoke check ──────────────────────────────────────────────────────

const cases = [
    { input: '1 + 2',            expected: 3  },
    { input: '2 + 3 * 4',        expected: 14 },
    { input: '(2 + 3) * 4',      expected: 20 },
    { input: '10 - 4 / 2',       expected: 8  },
];

for (const { input, expected } of cases) {
    const lexer = new Lexer(input);
    const got = parseExpr(lexer);
    if (got !== expected) {
        throw new Error(`mismatch: ${input} => ${got} !== ${expected}`);
    }
}

console.log(`${cases.length} cases passed`);

File: tests/corpus/concept.cpp
// vyakarana stand-in -- not from vidya. See docs/adr/0006-standin-corpus-policy.md.
// Re-sync when vidya adds a C++ reference sample.
//
// Lexing + parsing of arithmetic expressions, modern C++ idioms.

#include <cstdint>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>
#include <vector>

namespace vyakarana::concept {

// ── Tokens ───────────────────────────────────────────────────────────

enum class TokenKind {
    Number,
    Plus,
    Minus,
    Star,
    Slash,
    LParen,
    RParen,
    Eof,
};

struct Token {
    TokenKind kind;
    std::string text;
    std::size_t pos;
};

// ── Lexer ────────────────────────────────────────────────────────────

class Lexer {
public:
    explicit Lexer(std::string source) : source_(std::move(source)), pos_(0) {}

    Token next() {
        skipWhitespace();
        const std::size_t start = pos_;

        if (pos_ >= source_.size()) {
            return Token{TokenKind::Eof, "", pos_};
        }

        const char c = source_[pos_];
        switch (c) {
            case '+': ++pos_; return Token{TokenKind::Plus,   "+", start};
            case '-': ++pos_; return Token{TokenKind::Minus,  "-", start};
            case '*': ++pos_; return Token{TokenKind::Star,   "*", start};
            case '/': ++pos_; return Token{TokenKind::Slash,  "/", start};
            case '(': ++pos_; return Token{TokenKind::LParen, "(", start};
            case ')': ++pos_; return Token{TokenKind::RParen, ")", start};
            default: break;
        }

        if (isDigit(c)) {
            while (pos_ < source_.size() && isDigit(source_[pos_])) {
                ++pos_;
            }
            return Token{TokenKind::Number, source_.substr(start, pos_ - start), start};
        }

        throw std::runtime_error("unexpected character at " + std::to_string(pos_));
    }

private:
    static bool isDigit(char c) { return c >= '0' && c <= '9'; }

    void skipWhitespace() {
        while (pos_ < source_.size() && std::isspace(static_cast<unsigned char>(source_[pos_]))) {
            ++pos_;
        }
    }

    std::string source_;
    std::size_t pos_;
};

// ── AST ──────────────────────────────────────────────────────────────

struct Expr {
    virtual ~Expr() = default;
    virtual std::int64_t eval() const = 0;
};

using ExprPtr = std::unique_ptr<Expr>;

struct NumberExpr final : Expr {
    std::int64_t value;
    explicit NumberExpr(std::int64_t v) : value(v) {}
    std::int64_t eval() const override { return value; }
};

struct UnaryExpr final : Expr {
    char op;
    ExprPtr operand;
    UnaryExpr(char o, ExprPtr e) : op(o), operand(std::move(e)) {}
    std::int64_t eval() const override { return -operand->eval(); }
};

struct BinOpExpr final : Expr {
    char op;
    ExprPtr left;
    ExprPtr right;
    BinOpExpr(char o, ExprPtr l, ExprPtr r)
        : op(o), left(std::move(l)), right(std::move(r)) {}

    std::int64_t eval() const override {
        const auto a = left->eval();
        const auto b = right->eval();
        switch (op) {
            case '+': return a + b;
            case '-': return a - b;
            case '*': return a * b;
            case '/': return a / b;
            default:  throw std::runtime_error("bad op");
        }
    }
};

// ── Pratt parser ─────────────────────────────────────────────────────

class Parser {
public:
    explicit Parser(std::string source)
        : lexer_(std::move(source)), current_(lexer_.next()) {}

    ExprPtr parseExpr(int minBp) {
        ExprPtr left = parsePrimary();
        while (true) {
            const int bp = infixBp(current_.kind);
            if (bp == 0 || bp < minBp) {
                break;
            }
            const char op = current_.text[0];
            advance();
            ExprPtr right = parseExpr(bp + 1);
            left = std::make_unique<BinOpExpr>(op, std::move(left), std::move(right));
        }
        return left;
    }

private:
    void advance() { current_ = lexer_.next(); }

    static int infixBp(TokenKind k) {
        switch (k) {
            case TokenKind::Plus:
            case TokenKind::Minus: return 1;
            case TokenKind::Star:
            case TokenKind::Slash: return 3;
            default:               return 0;
        }
    }

    ExprPtr parsePrimary() {
        if (current_.kind == TokenKind::Minus) {
            advance();
            return std::make_unique<UnaryExpr>('-', parsePrimary());
        }
        if (current_.kind == TokenKind::LParen) {
            advance();
            ExprPtr inner = parseExpr(0);
            if (current_.kind != TokenKind::RParen) {
                throw std::runtime_error("expected )");
            }
            advance();
            return inner;
        }
        if (current_.kind == TokenKind::Number) {
            std::int64_t v = std::stoll(current_.text);
            advance();
            return std::make_unique<NumberExpr>(v);
        }
        throw std::runtime_error("unexpected token");
    }

    Lexer lexer_;
    Token current_;
};

}  // namespace vyakarana::concept

int main() {
    using vyakarana::concept::Parser;
    const std::vector<std::pair<std::string, std::int64_t>> cases = {
        {"1 + 2",       3},
        {"2 * 3 + 4",   10},
        {"(1 + 2) * 3", 9},
        {"-5 + 3",      -2},
    };
    for (const auto& [input, expected] : cases) {
        const auto got = Parser(input).parseExpr(0)->eval();
        std::cout << input << " => " << got << " (expected " << expected << ")\n";
        if (got != expected) {
            throw std::runtime_error("mismatch");
        }
    }
}

// 2.3.2 error-hole coverage: C++14 digit separators and a
// backslash line continuation in a macro. Both are in c.cyml
// already; cpp.cyml had simply diverged.
constexpr long kMaxTokens = 1'000'000;

#define TRACE_TOKEN(k, n) \
    do { (void)(k); (void)(n); } while (0)
